js/navigation/index.js

```diff
@@ -1,82 +1,73 @@
 // @flow
 
 import React, { Component } from 'react'
+import { BackHandler } from 'react-native'
 import { connect } from 'react-redux'
-import { StackNavigator, addNavigationHelpers } from 'react-navigation'
+import { addNavigationHelpers, NavigationActions } from 'react-navigation'
+import { createReduxBoundAddListener } from 'react-navigation-redux-helpers'
 
 import type { MapStateToProps } from 'react-redux'
-import type { NavigationState } from 'react-navigation/src/TypeDefinition'
+import type { NavigationState } from 'react-navigation'
 
-import ROUTES from './routes'
-import LoginScreen from '../components/LoginScreen'
-import ProfileScreen from '../components/ProfileScreen'
+import MainNavigator from './MainNavigator'
 
-import type { ApiUserProfile } from '../utils/api'
 import type { Dispatch } from '../actions/types'
 
-// Initialize the stack navigator
-const HomeRoutes = {
-  [ROUTES.HOME]: {
-    screen: LoginScreen
-  },
-
-  [ROUTES.PROFILE]: {
-    screen: ProfileScreen
-  }
-}
-
-export const HomeNavigator = StackNavigator(
-  {
-    ...HomeRoutes
-  },
-  {
-    initialRouteName: ROUTES.HOME,
-
-    // Let each screen handle the header and navigation
-    headerMode: 'none'
-  }
-)
-
-export const ProfileNavigator = StackNavigator(
-  {
-    ...HomeRoutes
-  },
-  {
-    initialRouteName: ROUTES.PROFILE,
-
-    // Let each screen handle the header and navigation
-    headerMode: 'none'
-  }
-)
+/**
+ * A listener of the new react-navigation redux middleware.
+ * The parameter must be the same used in createReactNavigationReduxMiddleware function.
+ */
+const addListener = createReduxBoundAddListener('root')
 
 type Props = {
-  profile: ApiUserProfile,
-  nav: NavigationState,
+  navigation: NavigationState,
   dispatch: Dispatch
 }
 
 /**
  * Main app navigator.
  */
 class Navigation extends Component<Props> {
-  render() {
-    const { profile } = this.props
-    const Navigator = profile ? ProfileNavigator : HomeNavigator
+  componentDidMount() {
+    // Add an handler for the hardware back button in Android
+    BackHandler.addEventListener('hardwareBackPress', this.onBackPress)
+  }
 
+  componentWillUnmount() {
+    // Remove handler for the hardware back button in Android
+    BackHandler.removeEventListener('hardwareBackPress', this.onBackPress)
+  }
+
+  /**
+   * Handle the hardware back button in Android.
+   * It returns a boolean that if true avoid invoking the default back button
+   * functionality to exit the app.
+   */
+  onBackPress = (): boolean => {
+    const { dispatch, navigation } = this.props
+    // If we are on the first screen of the stack we can exit from the application
```

Comment:
what does it mean exactly? is it going to close the application?

Reply:
Yes it is going to close the application.